contracts/governance/Governor.sol

```diff
@@ -4,208 +4,224 @@ pragma solidity ^0.8.0;
-    {
+    constructor(string memory name_) EIP712(name_, version()) {
         _name = name_;
-        _version = version_;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165, ERC165) returns (bool) {
+        return interfaceId == type(IGovernor).interfaceId || super.supportsInterface(interfaceId);
     }
 
     /**
      * @dev See {IGovernor-name}.
      */
-    function name() external view virtual override returns (string memory) {
+    function name() public view virtual override returns (string memory) {
         return _name;
     }
 
     /**
-     * @dev See {IGovernor-version}.
+     * @dev See {IGovernor-vestion}.
+     */
+    function version() public view virtual override returns (string memory) {
+        return "1";
+    }
+
+    /**
+     * @dev See {IGovernor-hashProposal}.
      */
-    function version() external view virtual override returns (string memory) {
-        return _version;
+    function hashProposal(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public pure virtual override returns (uint256) {
+        return uint256(keccak256(abi.encode(targets, values, calldatas, descriptionHash)));
     }
 
     /**
-     * @dev See {IGovernor-version}.
+     * @dev See {IGovernor-state}.
      */
     function state(uint256 proposalId) public view virtual override returns (ProposalState) {
-        Proposal memory proposal = _proposals[proposalId];
+        ProposalCore memory proposal = _proposals[proposalId];
 
-        if (proposal.timer.isUnset()) {
-            // There is no ProposalState for unset proposals
-            revert("Governor::state: invalid proposal id");
-        } else if (block.number <= proposal.snapshot) {
-            return ProposalState.Pending;
-        } else if (proposal.timer.isPending()) {
-            return ProposalState.Active;
-        } else if (proposal.timer.isExpired()) {
-            return (proposalWeight(proposalId) >= quorum(proposal.snapshot) && _voteSuccess(proposalId))
-                ? ProposalState.Succeeded
-                : ProposalState.Defeated;
+        if (proposal.executed) {
+            return ProposalState.Executed;
         } else if (proposal.canceled) {
             return ProposalState.Canceled;
+        } else if (proposal.voteStart.isPending()) {
+            return ProposalState.Pending;
+        } else if (proposal.voteEnd.isPending()) {
+            return ProposalState.Active;
+        } else if (proposal.voteEnd.isExpired()) {
+            return
+                _quorumReached(proposalId) && _voteSuccess(proposalId)
+                    ? ProposalState.Succeeded
+                    : ProposalState.Defeated;
         } else {
-            return ProposalState.Executed;
+            revert("Governor::state: invalid proposal id");
```

Comment:
I think we should use the revert reason convention we've used so far which would just be `Governor: invalid proposal id`.

I think "unknown" proposal id would be better by the way.

Reply:
I did follow this https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/GovernorAlpha.sol#L227
Revert message compatibility might not be important, in that case I'd be ok to use `Governor: unknown proposal id` instead

Reply:
I wouldn't worry about revert reason compatibility, we plan to change them to custom errors soon anyway.